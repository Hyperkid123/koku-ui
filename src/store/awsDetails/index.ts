import * as awsDetailsActions from './awsDetailsActions';
import {
  awsDetailsStateKey,
  AwsDetailsTab,
  AwsDetailsWidget,
} from './awsDetailsCommon';
import { awsDetailsReducer } from './awsDetailsReducer';
import * as awsDetailsSelectors from './awsDetailsSelectors';

export {
  awsDetailsStateKey,
  awsDetailsReducer,
  awsDetailsActions,
  awsDetailsSelectors,
  AwsDetailsTab,
  AwsDetailsWidget,
};
