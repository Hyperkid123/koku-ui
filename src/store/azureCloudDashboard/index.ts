import * as azureCloudDashboardActions from './azureCloudDashboardActions';
import {
  azureCloudDashboardStateKey,
  AzureCloudDashboardTab,
  AzureCloudDashboardWidget,
} from './azureCloudDashboardCommon';
import { azureCloudDashboardReducer } from './azureCloudDashboardReducer';
import * as azureCloudDashboardSelectors from './azureCloudDashboardSelectors';

export {
  azureCloudDashboardStateKey,
  azureCloudDashboardReducer,
  azureCloudDashboardActions,
  azureCloudDashboardSelectors,
  AzureCloudDashboardTab,
  AzureCloudDashboardWidget,
};
