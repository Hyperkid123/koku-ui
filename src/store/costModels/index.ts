import * as costModelsActions from './actions';
import {
  CostModelsAction,
  CostModelsState,
  reducer as costModelsReducer,
  stateKey as costModelsStateKey,
} from './reducer';
import * as costModelsSelectors from './selectors';

export {
  CostModelsAction,
  costModelsActions,
  costModelsReducer,
  costModelsSelectors,
  CostModelsState,
  costModelsStateKey,
};
