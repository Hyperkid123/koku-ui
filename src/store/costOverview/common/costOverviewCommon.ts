import { ReportPathsType, ReportType } from 'api/reports/report';

export const enum CostOverviewWidgetType {
  costBreakdown = 'costBreakdown', // This type displays cost breakdown pie chart
  cpuUsage = 'cpuUsage', // This type displays CPU usage
  memoryUsage = 'memoryUsage', // This type displays memory usage
  reportSummary = 'summary', // This type displays cost report summary
}

export interface CostOverviewWidget {
  id: number;
  costBreakdown?: {
    reportPathsType: ReportPathsType; // Report URL path
    reportType: ReportType; // Report type; cost, storage, etc.
  };
  cpuUsage?: {
    reportPathsType: ReportPathsType; // Report URL path
    reportType: ReportType; // Report type; cost, storage, etc.
    showCapacityOnGroupBy?: string[]; // Show capacity when group_by is matched
  };
  memoryUsage?: {
    reportPathsType: ReportPathsType; // Report URL path
    reportType: ReportType; // Report type; cost, storage, etc.
    showCapacityOnGroupBy?: string[]; // Show capacity when group_by is matched
  };
  reportSummary?: {
    reportGroupBy: string; // Report group_by
    reportPathsType: ReportPathsType; // Report URL path
    reportType: ReportType; // Report type; cost, storage, etc.
    showWidgetOnGroupBy?: string[]; // Show widget when group_by is matched
    usePlaceholder?: boolean; // Use placeholder to keep card placement when widget is not shown
  };
  type: CostOverviewWidgetType;
}
