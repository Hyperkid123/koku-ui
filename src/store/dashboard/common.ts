export interface ValueFormatOptions {
  fractionDigits?: number;
}

export interface DashboardWidget {
  availableTabs?: any[];
  currentTab: any;
  details: {
    costKey?: string /** i18n label key */;
    formatOptions: ValueFormatOptions;
    showUnits?: boolean;
    showUsageLegendLabel?: boolean;
    usageFormatOptions?: ValueFormatOptions;
    usageKey?: string /** i18n label key */;
  };
  filter?: {
    limit?: number;
    service?: string;
  };
  id: number;
  isDetailsLink?: boolean;
  isHorizontal?: boolean;
  isUsageFirst?: boolean;
  reportType: any;
  /** i18n key for the title. passed { startDate, endDate, month, time } */
  titleKey: string;
  tabsFilter?: {
    limit?: number;
    service?: string;
  };
  trend: {
    titleKey: string;
    type: number;
    formatOptions: ValueFormatOptions;
  };
  topItems: {
    formatOptions: {};
  };
}
