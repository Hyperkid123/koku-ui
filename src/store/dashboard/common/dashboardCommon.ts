export const enum DashboardChartType {
  cost = 'cost',
  trend = 'trend',
  usage = 'usage',
}

export const enum DashboardPerspective {
  aws = 'aws',
  awsCloud = 'aws_cloud', // Aws filtered by Ocp
  azure = 'azure',
  azureCloud = 'azure_cloud', // Azure filtered by Ocp
  ocp = 'ocp',
  ocpCloud = 'ocp_cloud', // All filtered by Ocp
  ocpUsage = 'ocp_usage',
  ocpSupplementary = 'ocp_supplementary',
}

export interface ValueFormatOptions {
  fractionDigits?: number;
}

export interface DashboardWidget {
  availableTabs?: any[];
  chartType?: DashboardChartType;
  currentTab: any;
  details: {
    costKey?: string /** i18n label key */;
    formatOptions: ValueFormatOptions;
    showUnits?: boolean;
    showUsageLegendLabel?: boolean;
    units?: string;
    usageFormatOptions?: ValueFormatOptions;
    usageKey?: string /** i18n label key */;
  };
  filter?: {
    limit?: number;
    service?: string;
    service_name?: string;
  };
  id: number;
  isDetailsLink?: boolean;
  isHorizontal?: boolean;
  isUsageFirst?: boolean;
  perspective?: DashboardPerspective;
  reportType: any;
  /** i18n key for the title. passed { startDate, endDate, month, time } */
  titleKey: string;
  tabsFilter?: {
    limit?: number;
    service?: string;
    service_name?: string;
  };
  trend: {
    titleKey: string;
    type: number;
    formatOptions: ValueFormatOptions;
  };
  topItems: {
    formatOptions: {};
  };
}
