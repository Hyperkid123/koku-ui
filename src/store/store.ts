import { notificationsMiddleware } from '@red-hat-insights/insights-frontend-components/components/Notifications';
import axios from 'axios';
import { applyMiddleware, compose, createStore, DeepPartial } from 'redux';
import thunk from 'redux-thunk';
import { rootReducer, RootState } from './rootReducer';
import { sessionActions } from './session';

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__(options: any): typeof compose;
  }
}

const logger = () => store => next => action => {
  return next(action);
};

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ serialize: true })
    : compose;

export const middlewares = [
  logger(),
  thunk,
  notificationsMiddleware(),
  logger(),
];

export function configureStore(initialState: DeepPartial<RootState>) {
  const enhancer = composeEnhancers(applyMiddleware(...middlewares));
  const store = createStore(rootReducer, initialState, enhancer);

  axios.interceptors.response.use(null, error => {
    if (
      error &&
      error.response &&
      (error.response.status === 403 || error.response.status === 401)
    ) {
      store.dispatch(sessionActions.logout());
    }
    return Promise.reject(error);
  });

  return store;
}
