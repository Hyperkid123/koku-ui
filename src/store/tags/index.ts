import * as tagActions from './tagActions';
import { tagStateKey } from './tagCommon';
import { CachedTag, TagAction, tagReducer, TagState } from './tagReducer';
import * as tagSelectors from './tagSelectors';

export { TagAction, CachedTag, tagActions, tagReducer, tagSelectors, TagState, tagStateKey };
