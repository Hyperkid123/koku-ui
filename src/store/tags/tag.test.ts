jest.mock('api/tags/tagUtils');

import { waitFor } from '@testing-library/react';
import type { Tag } from 'api/tags/tag';
import { TagPathsType, TagType } from 'api/tags/tag';
import { runTag } from 'api/tags/tagUtils';
import { FetchStatus } from 'store/common';
import { createMockStoreCreator } from 'store/mockStore';

import * as actions from './tagActions';
import { tagStateKey } from './tagCommon';
import { tagReducer } from './tagReducer';
import * as selectors from './tagSelectors';

const createTagsStore = createMockStoreCreator({
  [tagStateKey]: tagReducer,
});

const runTagMock = runTag as jest.Mock;

const mockTagReport: Tag = {
  data: [],
  total: {
    value: 100,
    units: 'USD',
  },
} as any;

const tagReportType = TagType.tag;
const tagReportPathsType = TagPathsType.aws;
const tagQueryString = 'tagQueryString';

runTagMock.mockResolvedValue({ data: mockTagReport });
global.Date.now = jest.fn(() => 12345);

jest.spyOn(actions, 'fetchTag');
jest.spyOn(selectors, 'selectTagFetchStatus');

test('default state', () => {
  const store = createTagsStore();
  expect(selectors.selectTagState(store.getState())).toMatchSnapshot();
});

test('fetch tag report success', async () => {
  const store = createTagsStore();
  store.dispatch(actions.fetchTag(tagReportPathsType, tagReportType, tagQueryString));
  expect(runTagMock).toBeCalled();
  expect(selectors.selectTagFetchStatus(store.getState(), tagReportPathsType, tagReportType, tagQueryString)).toBe(
    FetchStatus.inProgress
  );
  await waitFor(() => expect(selectors.selectTagFetchStatus).toHaveBeenCalled());
  const finishedState = store.getState();
  expect(selectors.selectTagFetchStatus(finishedState, tagReportPathsType, tagReportType, tagQueryString)).toBe(
    FetchStatus.complete
  );
  expect(selectors.selectTagError(finishedState, tagReportPathsType, tagReportType, tagQueryString)).toBe(null);
});

test('fetch tag report failure', async () => {
  const store = createTagsStore();
  const error = Symbol('tag error');
  runTagMock.mockRejectedValueOnce(error);
  store.dispatch(actions.fetchTag(tagReportPathsType, tagReportType, tagQueryString));
  expect(runTag).toBeCalled();
  expect(selectors.selectTagFetchStatus(store.getState(), tagReportPathsType, tagReportType, tagQueryString)).toBe(
    FetchStatus.inProgress
  );
  await waitFor(() => expect(selectors.selectTagFetchStatus).toHaveBeenCalled());
  const finishedState = store.getState();
  expect(selectors.selectTagFetchStatus(finishedState, tagReportPathsType, tagReportType, tagQueryString)).toBe(
    FetchStatus.complete
  );
  expect(selectors.selectTagError(finishedState, tagReportPathsType, tagReportType, tagQueryString)).toBe(error);
});

test('does not fetch tag report if the request is in progress', () => {
  const store = createTagsStore();
  store.dispatch(actions.fetchTag(tagReportPathsType, tagReportType, tagQueryString));
  store.dispatch(actions.fetchTag(tagReportPathsType, tagReportType, tagQueryString));
  expect(runTag).toHaveBeenCalledTimes(1);
});

test('tag report is not refetched if it has not expired', async () => {
  const store = createTagsStore();
  store.dispatch(actions.fetchTag(tagReportPathsType, tagReportType, tagQueryString));
  await waitFor(() => expect(actions.fetchTag).toHaveBeenCalled());
  store.dispatch(actions.fetchTag(tagReportPathsType, tagReportType, tagQueryString));
  expect(runTag).toHaveBeenCalledTimes(1);
});
