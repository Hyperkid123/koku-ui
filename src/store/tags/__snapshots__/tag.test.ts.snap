// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default state 1`] = `
Object {
  "byId": Map {},
  "errors": Map {},
  "fetchStatus": Map {},
}
`;

exports[`fetch tag report success 1`] = `
Object {
  "data": Array [],
  "timeRequested": 1633111749670,
  "total": Object {
    "units": "USD",
    "value": 100,
  },
}
`;
